docs/syntax.rst

```diff
@@ -29,102 +25,71 @@ The return value will be automatically casted into the hinted type. e.g.,
 
 .. note::
 
-    For now, we only support one scalar as return value. Returning ``ti.Matrix`` or `ti.Vector`` is not supported. Python-style tuple return is not supported. e.g.:
+    For now, we only support one scalar as return value. Returning ``ti.Matrix`` or ``ti.Vector`` is not supported. Python-style tuple return is not supported either. For example,
 
     .. code-block:: python
 
         @ti.kernel
         def bad_kernel() -> ti.Matrix:
-            return ti.Matrix([[1, 0], [0, 1]])  # ERROR!
+            return ti.Matrix([[1, 0], [0, 1]])  # Error
 
         @ti.kernel
         def bad_kernel() -> (ti.i32, ti.f32):
             x = 1
             y = 0.5
-            return x, y  #  ERROR!
-
-.. note::
-  For correct gradient behaviors in differentiable programming, please refrain from using kernel return values. Instead, store the result into a global variable (e.g. ``loss[None]``).
-
-(TODO: move the following to advanced topics)
-
-* For differentiable programming kernels should better have either serial statements or a single parallel for-loop. If you don't use differentiable programming, feel free to ignore this tip.
-
-.. code-block:: python
-
-        @ti.kernel
-        def a_hard_kernel_to_auto_differentiate():
-          sum = 0
-          for i in x:
-            sum += x[i]
-          for i in y:
-            y[i] = sum
+            return x, y  # Error
 
-        # instead, split it into multiple kernels to be nice to the Taichi autodiff compiler:
 
-        @ti.kernel
-        def reduce():
-          for i in x:
-            sum[None] += x[i]
+We also support **template arguments** (see :ref:`template_metaprogramming`) and **external array arguments** (see :ref:`external`) in Taichi kernels.
 
-        @ti.kernel
-        def assign()
-          for i in y:
-            y[i] = sum[None]
+.. warning::
 
-        def main():
-          with ti.Tape(loss):
-            ...
-            sum[None] = 0
-            reduce()
-            assign()
-            ...
+   When using differentiable programming, there are a few more constraints on kernel structures. See the **Kernel Simplicity Rule** in :ref:`differentiable`.
 
+   Also, please do not use kernel return values in differentiable programming, since the return value will not be tracked by automatic differentiation. Instead, store the result into a global variable (e.g. ``loss[None]``).
 
 Functions
------------------------------------------------
+---------
 
-Use ``@ti.func`` to decorate your Taichi functions. These functions are callable only in `Taichi`-scope. Don't call them in `Python`-scope. All function calls are force-inlined, so no recursion supported.
+Use ``@ti.func`` to decorate your Taichi functions. These functions are callable only in `Taichi`-scope. Do not call them in `Python`-scopes.
```

Comment:
```suggestion
Use ``@ti.func`` to decorate your Taichi functions. These functions are callable only in `Taichi`-scope. Do not call them in `Python`-scope.
```
Let's **always keep singular form**, unless it's clarified by context. I guess that's just common for English to lack a netural version for a word, like `man` and `woman`, and you have to choose one of them to stand for both...

Reply:
I feel like we should always use `scopes`. The singular form must be `a scope` or `the scope` in a sentence. We should use plural forms consistently since we are not talking about a specific scope.

Reply:
Actually, either option is fine as long as we use it consistently across the documentation.

Reply:
```suggestion
Use ``@ti.func`` to decorate your Taichi functions. These functions are callable only in `Taichi`-scopes. Do not call them in `Python`-scopes.
```

Reply:
How about this: given there are many places in the document where `Taichi-scope` is already used, let's use the singular form as you suggested throughout the documentation. In a separate PR I'll do a pass to unify all usage of `scope`.

Reply:
Ack :)